bazel/grpc_build_system.bzl

```diff
@@ -298,39 +310,58 @@ def grpc_cc_test(name, srcs = [], deps = [], external_deps = [], args = [], data
         "linkstatic": linkstatic,
     }
     if uses_polling:
-        # the vanilla version of the test should run on platforms that only
-        # support a single poller
-        native.cc_test(
-            name = name,
-            testonly = True,
-            tags = (tags + [
-                "no_linux",  # linux supports multiple pollers
-            ]),
-            **args
-        )
-
-        # on linux we run the same test multiple times, once for each poller
-        for poller in POLLERS:
-            native.sh_test(
-                name = name + "@poller=" + poller,
-                data = [name] + data,
-                srcs = [
-                    "//test/core/util:run_with_poller_sh",
-                ],
-                size = size,
-                timeout = timeout,
-                args = [
-                    poller,
-                    "$(location %s)" % name,
-                ] + args["args"],
-                tags = (tags + ["no_windows", "no_mac"]),
-                exec_compatible_with = exec_compatible_with,
-                exec_properties = exec_properties,
-                shard_count = shard_count,
-                flaky = flaky,
-            )
+        # When true, EventEngines will only be exercised against one poller
+        engine_only = False
+        for engine in EVENT_ENGINES:
+            if "no_windows" not in engine["tags"] and "no_mac" not in engine["tags"]:
+                # the vanilla version of the test should run on platforms that only
```

Comment:
The logic flow here is fairly complex and hard to follow.  Can we simplify this somehow?

For example, maybe we can generate the basic test rule as a `cc_library` instead of a `cc_test`, and then generate separate test targets for each platform that depend on that `cc_library`?  I'm thinking of something like this (not tested; there are probably some details that need to be changed):

```
# cc_library target for test code itself
native.cc_library(
    name = name + "_lib",
    testonly = True,
    **args
)

# Non-Linux platforms.
if "no_windows" not in engine["tags"] and "no_mac" not in engine["tags"]:
  # These platforms do not support multiple polling engines,
  # so just create a target for each EventEngine.
  for engine in EVENT_ENGINE:
    native.cc_test(
      name = name + "@engine=" + engine["name"],
      deps = deps + (name + "_lib"),
      tags = tags + engine["tags"] + ["no_linux"],
      env = {"GRPC_EVENTENGINE_STRATEGY": engine["name"]},
      **args
    )

# Linux platforms.
if "no_linux" not in engine["tags"]:
  # Generate one test for each poller, all using the first EventEngine.
  for poller in POLLERS:
    native.cc_test(
      name = name + "@poller=" + poller + ",engine=" + EVENT_ENGINES[0]["name"],
      deps = deps + (name + "_lib"),
      tags = tags + EVENT_ENGINES[0]["tags"] + ["no_windows", "no_mac"]),
      # ...etc...
    )
  # Now generate one test for each subsequent EventEngine,
  # all using the first poller.
  for engine in EVENT_ENGINES[1:]:
    native.cc_test(
      name = name + "@poller=" + POLLERS[0] + ",engine=" + engine["name"],
      deps = deps + (name + "_lib"),
      tags = tags + engine["tags"] + ["no_windows", "no_mac"]),
      # ...etc...
    )
```

Reply:
Sure, I've added back in some code duplication (for now) to make the logic easier to follow. The logic flows much like your pseudocode above, and generates the same results.

I'm on the fence about how far to dig in here to refactor the `native.sh_test`s into `native.cc_test`s as you coded above. I'd eventually like to deduplicate this file and `test/cpp/end2end/generate_tests.bzl`, which apparently uses a mostly-copied `native.sh_test` setup from here. But I'd like to handle that in a separate cleanup, if that's alright.